### ISSUE TITLE
`Sequence` disables typechecking?

### ISSUE THREAD

**Issue by bukzor-sentryio** [microsoft/pyright#5215]
The below code typechecks although it shouldn't. Swapping `Sequence` with `List` allows it to correctly fail.

```python
NestedIntList = Union[int, Sequence["NestedIntList"]]

def complex_interpreter(src: str) -> NestedIntList:
    nesting = src.count("[")
    result: NestedIntList = src.replace("[", "")
    for _ in range(nesting):
        result = [result]
    return result
```

```sh
$ pyright demo6.py
0 errors, 0 warnings, 0 informations

$ mypy demo6.py
Success: no issues found in 1 source file

$ pytype demo6.py
Success: no errors found
```

Full context:  https://github.com/microsoft/pyright/discussions/5214#discussioncomment-6053439

**Comment by erictraut** [comment]
Thanks for filing the bug. The issue is specific to the use of `Sequence` in a recursive type alias and the assignment of a `str` value. The `str` class derives from `Sequence[str]`, which is also a recursive type. In this case, the type checker is recursively checking to see if the two types (both of which are recursive Sequences) are type compatible. Without special handling, this would result in infinite recursion and a stack overflow. Pyright therefore contains some internal limits that detect this case. When these limits are hit, it gives up and conservatively guesses that the types are compatible. This is the correct guess in most cases, but in this case it's incorrect. I don't see a way to fix this edge case without breaking common patterns or introducing some very complex and expensive logic. 

I'll give this a bit more thought to see if I can come up with a solution, but my current inclination is to close this as "won't fix".

**Comment by bukzor-sentryio** [comment]
> conservatively guesses that the types are compatible

that seems a liberal rather than conservative guess, to me :)

I for one would make heavy use of an ability to make my `str`s no longer `Sequence`s. Perhaps a `AtomicStr` in type_extensions? It may be that it already exists... but on a quick scan I don't find any.

Before you close wontfix, could I at least trouble you to show (or link to pertinent reading) the common counterexamples? It would put me at ease to be able to see that there's no good way out (especially when I come back to this issue ~18mo later).

**Comment by erictraut** [comment]
> that seems a liberal rather than conservative guess, to me :)

It's conservative in the sense that it tries to avoid false positives (potentially at the expense of false negatives). That's a general principle we use throughout pyright. All other considerations being equal, we tend to eliminate false positives.

**Comment by erictraut** [comment]
This will be addressed in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.312, which I just published. It will also be included in the next insiders release of pylance.
